notify/src/address/tracker.rs

```diff
@@ -0,0 +1,649 @@
+        ((max_addresses + 1) * 8 / 7).next_power_of_two() * 7 / 8 - 1
+    }
+
+    fn new(max_addresses: Option<usize>) -> Self {
+        // Expand the maximum address count to the IndexMap actual usable allocated size minus 1.
+        // Saving one entry for the insert/swap_remove scheme during entry recycling prevents a reallocation
+        // when reaching the maximum.
+        let max_addresses = max_addresses.map(Self::expand_max_addresses);
+        let capacity = max_addresses.map(|x| x + 1).unwrap_or_default();
+        let script_pub_keys = IndexMap::with_capacity(capacity);
+        debug!("Creating an address tracker with a capacity of {}", script_pub_keys.capacity());
+        if let Some(max_addresses) = max_addresses {
+            info!("Tracking UTXO changed events for {} addresses at most", max_addresses);
+        }
+        let empty_entries = HashSet::with_capacity(capacity);
+        Self { script_pub_keys, max_addresses, empty_entries }
+    }
+
+    fn is_full(&self) -> bool {
+        match self.max_addresses {
+            Some(max_addresses) => self.script_pub_keys.len() >= max_addresses && self.empty_entries.is_empty(),
+            None => false,
+        }
+    }
+
+    fn get(&self, spk: &ScriptPublicKey) -> Option<(Index, RefCount)> {
+        self.script_pub_keys.get_full(spk).map(|(index, _, count)| (index as Index, *count))
+    }
+
+    fn get_index(&self, index: Index) -> Option<&ScriptPublicKey> {
+        self.script_pub_keys.get_index(index as usize).map(|(spk, _)| spk)
+    }
+
+    fn get_index_address(&self, index: Index, prefix: Prefix) -> Option<Address> {
+        self.script_pub_keys
+            .get_index(index as usize)
+            .map(|(spk, _)| extract_script_pub_key_address(spk, prefix).expect("is retro-convertible"))
+    }
+
+    fn get_or_insert(&mut self, spk: ScriptPublicKey) -> Result<Index> {
+        match self.is_full() {
+            false => match self.script_pub_keys.entry(spk) {
+                Entry::Occupied(entry) => Ok(entry.index() as Index),
+                Entry::Vacant(entry) => {
+                    let mut index = entry.index() as Index;
+                    trace!(
+                        "AddressTracker insert #{} {}",
+                        index,
+                        extract_script_pub_key_address(entry.key(), Prefix::Mainnet).unwrap()
+                    );
+                    let _ = *entry.insert(0);
+
+                    // Recycle empty entries
+                    let mut recycled = false;
+                    if (index + 1) as usize == self.script_pub_keys.len() && !self.empty_entries.is_empty() {
+                        let empty_index = self.empty_entries.iter().cloned().next();
+                        if let Some(empty_index) = empty_index {
+                            self.script_pub_keys.swap_remove_index(empty_index as usize);
+                            index = empty_index;
+                            recycled = true;
```

Comment:
It took me a while to figure out that `empty_index` is removed from `self.empty_entries` only when the count is increased via `inc_count`. Perhaps a comment here would help